internal/datacoord/sync_segments_scheduler.go

```diff
@@ -0,0 +1,155 @@
+
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/logutil"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+type SyncSegmentsScheduler struct {
+	quit chan struct{}
+	wg   sync.WaitGroup
+
+	meta           *meta
+	channelManager ChannelManager
+	sessions       SessionManager
+}
+
+func newSyncSegmentsScheduler(m *meta, channelManager ChannelManager, sessions SessionManager) *SyncSegmentsScheduler {
+	return &SyncSegmentsScheduler{
+		quit:           make(chan struct{}),
+		wg:             sync.WaitGroup{},
+		meta:           m,
+		channelManager: channelManager,
+		sessions:       sessions,
+	}
+}
+
+func (sss *SyncSegmentsScheduler) Start() {
+	sss.quit = make(chan struct{})
+	ticker := time.NewTicker(Params.DataCoordCfg.SyncSegmentsInterval.GetAsDuration(time.Second))
+	sss.wg.Add(1)
+
+	go func() {
+		defer logutil.LogPanic()
+		defer sss.wg.Done()
+
+		for {
+			select {
+			case <-sss.quit:
+				log.Info("sync segments scheduler quit")
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				sss.SyncSegmentsForCollections()
+			}
+		}
+	}()
+	log.Info("SyncSegmentsScheduler started...")
+}
+
+func (sss *SyncSegmentsScheduler) Stop() {
+	close(sss.quit)
+	sss.wg.Wait()
+}
+
+func (sss *SyncSegmentsScheduler) SyncSegmentsForCollections() {
+	collIDs := sss.meta.ListCollections()
+	for _, collID := range collIDs {
+		collInfo := sss.meta.GetCollection(collID)
+		if collInfo == nil {
+			log.Warn("collection info is nil, skip it", zap.Int64("collectionID", collID))
+			continue
+		}
+		pkField, err := typeutil.GetPrimaryFieldSchema(collInfo.Schema)
+		if err != nil {
+			log.Warn("get primary field from schema failed", zap.Int64("collectionID", collID),
+				zap.Error(err))
+			continue
+		}
+		for _, channelName := range collInfo.VChannelNames {
+			nodeID, err := sss.channelManager.FindWatcher(channelName)
+			if err != nil {
+				log.Warn("find watcher for channel failed", zap.Int64("collectionID", collID),
+					zap.String("channelName", channelName), zap.Error(err))
+				continue
+			}
+			for _, partitionID := range collInfo.Partitions {
+				if err := sss.SyncSegments(collID, partitionID, channelName, nodeID, pkField.GetFieldID()); err != nil {
+					log.Warn("sync segment with channel failed, retry next ticker",
+						zap.Int64("collectionID", collID),
+						zap.Int64("partitionID", partitionID),
+						zap.String("channel", channelName),
+						zap.Error(err))
+					continue
+				}
+			}
+		}
+	}
+}
+
+func (sss *SyncSegmentsScheduler) SyncSegments(collectionID, partitionID int64, channelName string, nodeID, pkFieldID int64) error {
+	log := log.With(zap.Int64("collectionID", collectionID), zap.Int64("partitionID", partitionID),
+		zap.String("channelName", channelName), zap.Int64("nodeID", nodeID))
+	segments := sss.meta.SelectSegments(WithCollection(collectionID), WithChannel(channelName), SegmentFilterFunc(func(info *SegmentInfo) bool {
```

Comment:
It seems unnecessary to filter again with WithCollection(collectionID) after already filtering with WithChannel(channelName).

Reply:
right, refine it.